Makefile

```diff
@@ -1,95 +1,155 @@
+############################################################
+# check section
+############################################################
+check: fmt vet lint
 
-# Run go fmt against code
-fmt:
-	go fmt ./pkg/... ./cmd/...
+fmt: depend generate
+	hack/verify-gofmt.sh
 
-# Run go vet against code
-vet:
-	go vet ./pkg/... ./cmd/...
+lint: depend generate
+ifndef HAS_LINT
+		go get -u golang.org/x/lint/golint
+		echo "installing golint"
+endif
+	hack/verify-golint.sh
 
-generate_yaml_test:
+vet: depend generate
+	go vet ./...
+
+############################################################
+# test section
+############################################################
+test: unit functional fmt vet generate_yaml_test
+
+unit: depend generate check
+	go test -tags=unit $(shell go list ./...) $(TESTARGS)
 
+functional:
+	@echo "$@ not yet implemented"
+
+# Generate manifests e.g. CRD, RBAC etc.
+generate_yaml_test:
 ifndef HAS_KUSTOMIZE
 	# for now, higher version has some problem so we stick to 1.0.11
 	wget https://github.com/kubernetes-sigs/kustomize/releases/download/v1.0.11/kustomize_1.0.11_linux_amd64
 	mv kustomize_1.0.11_linux_amd64 /usr/local/bin/kustomize
 	chmod +x /usr/local/bin/kustomize
 endif
-
 	# Create a dummy file for test only
 	echo 'clouds' > cmd/clusterctl/examples/ibmcloud/dummy-clouds-test.yaml
 	$(GENERATE_YAML_PATH)/$(GENERATE_YAML_EXEC) -f dummy-clouds-test.yaml ubuntu $(GENERATE_YAML_TEST_FOLDER)
 	# the folder will be generated under same folder of $(GENERATE_YAML_PATH)
 	rm -fr $(GENERATE_YAML_PATH)/$(GENERATE_YAML_TEST_FOLDER)
 	rm -f cmd/clusterctl/examples/ibmcloud/dummy-clouds-test.yaml
 
-# Generate code
-generate:
-ifndef GOPATH
-	$(error GOPATH not defined, please define GOPATH. Run "go help gopath" to learn more about GOPATH)
-endif
-	go generate ./pkg/... ./cmd/...
+############################################################
+# build section
+############################################################
+build: manager clusterctl
+
+manager: depend generate check
+	CGO_ENABLED=0 GOOS=$(GOOS) go build \
+		-ldflags $(LDFLAGS) \
+		-o bin/manager \
+		cmd/manager/main.go
+
+clusterctl: depend generate check
+	CGO_ENABLED=0 GOOS=$(GOOS) go build \
+		-ldflags $(LDFLAGS) \
+		-o bin/clusterctl \
+		cmd/clusterctl/main.go
+
+############################################################
+# deploy section
+############################################################
+# Run against the configured Kubernetes cluster in ~/.kube/config
+run: depend generate check
+	go run ./cmd/manager/main.go
+
+# Install CRDs into a cluster
+install: generate_yaml_test
+	kubectl apply -f config/crds
+
+# Deploy controller in the configured Kubernetes cluster in ~/.kube/config
+deploy: generate_yaml_test
+	cat provider-components.yaml | kubectl apply -f -
+
+############################################################
+# images section
+############################################################
+images: build docker-build
```

Comment:
Why `images` need `build`?

Can we rename `docker-build` to `build-images` directly and use it in test-infra?

s/docker-push/push-images/
s/quick-image/build-push-images/

Reply:
Oh, I see, we use docker multi-stage build in Dockerfile, https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/master/cmd/manager/Dockerfile#L17

So image building will not depend on external binary building, Ok to remove `build` dependency from `images`